File: app/tools/monitor.py
# -*- coding: utf-8 -*-
# 导入日期时间库
import datetime
# 导入系统信息库
import psutil


# 定义一个专门用于获取系统信息的类
class Monitor(object):
    # 专门用于单位转化的方法
    @staticmethod
    def bytes_to_gb(data, key=""):
        if key == "percent":
            # 百分比直接返回
            return data
        else:
            # 字节转化为GB，保留两位小数
            return round(data / (1024 ** 3), 2)

    # 专门获取CPU信息
    @staticmethod
    def cpu():
        # percpu：True获取每个CPU的使用率，False获取平均使用率
        data = dict(
            percent_avg=psutil.cpu_percent(interval=0, percpu=False),  # 平均使用率
            percent_per=psutil.cpu_percent(interval=0, percpu=True),  # 单独使用率
            num_p=psutil.cpu_count(logical=False),  # 物理CPU核心数
            num_l=psutil.cpu_count(logical=True)  # 逻辑CPU核心数
        )
        return data

    # 专门获取内存信息
    def mem(self):
        # 内存信息
        info = psutil.virtual_memory()
        data = dict(
            total=self.bytes_to_gb(info.total),  # 内存总量
            used=self.bytes_to_gb(info.used),  # 内存使用量
            free=self.bytes_to_gb(info.free),  # 内存剩余量
            percent=info.percent  # 内存使用率
        )
        return data

    # 专门获取交换分区/文件信息
    def swap(self):
        # 交换文件/分区信息
        info = psutil.swap_memory()
        data = dict(
            total=self.bytes_to_gb(info.total),  # 交换分区总量
            free=self.bytes_to_gb(info.free),  # 交换分区剩余量
            used=self.bytes_to_gb(info.used),  # 交换分区使用量
            percent=info.percent  # 交换分区使用率
        )
        return data

    # 专门获取磁盘信息
    def disk(self):
        # 专门获取磁盘分区信息
        info = psutil.disk_partitions()
        # 列表推导式
        data = [
            dict(
                device=v.device,  # 设备名称
                mountpoint=v.mountpoint,  # 挂载点
                fstype=v.fstype,  # 文件系统类型
                opts=v.opts,  # 操作选项
                used={
                    key: self.bytes_to_gb(val, key)
                    for key, val in psutil.disk_usage(v.mountpoint)._asdict().items()
                }  # 使用情况
            )
            for v in info if 'rw' in v.opts
        ]
        # if v.opts == 'rw,fixed'  # 获取可读可写的固定磁盘
        return data

    # 专门获取网络信息
    @staticmethod
    def net():
        # 获取地址信息
        address = psutil.net_if_addrs()
        # 处理地址信息
        address_info = {
            k: [
                dict(
                    family=val.family.name,  # 协议地址族名称
                    address=val.address,  # IP地址
                    netmask=val.netmask,  # 子网掩码
                    broadcast=val.broadcast  # 广播地址
                )
                for val in v if val.family.name == "AF_INET"  # 过滤IPV4协议
            ]
            for k, v in address.items()
        }
        # 获取网络信息
        io = psutil.net_io_counters(pernic=True)
        # 处理网络信息
        data = [
            dict(
                name=k,  # 网卡名称
                bytes_sent=v.bytes_sent,  # 发送字节数
                bytes_recv=v.bytes_recv,  # 接收字节数
                packets_sent=v.packets_sent,  # 发送包数
                packets_recv=v.packets_recv,  # 接收包数
                **address_info[k][0]
            )
            for k, v in io.items() if len(address_info[k]) > 0
        ]
        return data

    # 时间戳转化为时间字符方法
    @staticmethod
    def td(tm):
        dt = datetime.datetime.fromtimestamp(tm)
        return dt.strftime("%Y-%m-%d %H:%M:%S")

    # 获取日期时间
    @staticmethod
    def dt():
        return datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")

    # 专门获取最近开机时间
    def lasted_start_time(self):
        # 时间戳
        return self.td(psutil.boot_time())

    # 专门获取系统登录用户
    def login_users(self):
        # 获取系统登录用户
        users = psutil.users()
        # 处理系统登录用户
        data = [
            dict(
                name=v.name,  # 登录用户
                terminal=v.terminal,  # 登录终端
                host=v.host,  # 登录主机
                started=self.td(v.started),  # 登录时间
                pid=v.pid  # 进程ID
            )
            for v in users
        ]
        return data

File: app/static/js/monitor.js
if (!window.WebSocket) {
    window.WebSocket = window.MozWebSocket;
}

//1.定义长连接
let conn = null;

//进度条变化
function progress_status(val) {
    let data = "";
    if (val >= 0 && val < 25) {
        data = " bg-success";
    } else if (val >= 25 && val < 50) {
        data = "";
    } else if (val >= 50 && val < 75) {
        data = " bg-warning";
    } else if (val >= 75 && val <= 100) {
        data = " bg-success";
    }
    return data
}

//连接提示信息
function log(cls, msg) {
    document.getElementById("monitor_status").innerHTML = "<div class='alert alert-" + cls + "'>" + msg + "</div>";
}

//专门实时更新信息
function update_ui(e) {
    let data = e.data;
    data = JSON.parse(data); //把json字符串转化为对象
    /*平均CPU*/
    option_cpu_avg.series[0].data[0] = (data['cpu']['percent_avg'] / 100).toFixed(4);
    option_cpu_avg.title[0].text = data['dt'] + "-CPU平均使用率";
    chart_cpu_avg.setOption(option_cpu_avg);
    /*单独CPU*/
    let cpu_per = "";
    for (let k in data['cpu']['percent_per']) {
        let num = parseInt(k);
        cpu_per += "<tr><td class='text-primary' style='width: 30%'>CPU" + num + "</td>";
        cpu_per += "<td><div class='progress'><div class='progress-bar progress-bar-striped progress-bar-animated" + progress_status(data['cpu']['percent_per'][k]) + "' role='progressbar' aria-valuenow='" + data['cpu']['percent_per'][k] + "' aria-valuemin='0' aria-valuemax='100' style='width: " + data['cpu']['percent_per'][k] + "%'>" + data['cpu']['percent_per'][k] + "%</div></div></td></tr>";
    }
    document.getElementById("tb_cpu_per").innerHTML = cpu_per;
    /*内存实时更新*/
    option_mem.series[0].data[0].value = data['mem']['percent'];
    option_mem.title[0].text = data["dt"] + "-内存使用率";
    chart_mem.setOption(option_mem);
    document.getElementById("mem_percent").innerText = data['mem']['percent'];
    document.getElementById("mem_total").innerText = data['mem']['total'];
    document.getElementById("mem_used").innerText = data['mem']['used'];
    document.getElementById("mem_free").innerText = data['mem']['free'];

    /*交换分区实时更新*/
    option_swap.series[0].data[0].value = data['swap']['percent'];
    option_swap.title[0].text = data["dt"] + "-交换分区使用率";
    chart_swap.setOption(option_swap);
    document.getElementById("swap_percent").innerText = data['swap']['percent'];
    document.getElementById("swap_total").innerText = data['swap']['total'];
    document.getElementById("swap_used").innerText = data['swap']['used'];
    document.getElementById("swap_free").innerText = data['swap']['free'];

    /*网络信息更新*/
    let net = "";
    for (let k in data['net']) {
        let cd = data['net'][k];
        if (parseInt(cd['bytes_sent']) != 0 && parseInt(cd["bytes_recv"]) != 0) {
            let index = parseInt(k) + 1;
            let op = eval("option_net" + index);
            let ch = eval("chart_net" + index);
            op.title[0].text = data["dt"] + "-" + cd["name"] + "网卡信息";
            op.series[0].data = [
                {"name": "收包数", "value": cd["packets_recv"]},
                {"name": "发包数", "value": cd["packets_sent"]}
            ];
            op.series[1].data = [
                {"name": "收字节", "value": cd["bytes_recv"]},
                {"name": "发字节", "value": cd["bytes_sent"]}
            ];
            ch.setOption(op);
        }
        net += "<tr><td>" + cd['name'] + "</td>";
        net += "<td class='text-danger'>" + cd['bytes_sent'] + "</td>";
        net += "<td class='text-danger'>" + cd['bytes_recv'] + "</td>";
        net += "<td class='text-danger'>" + cd['packets_sent'] + "</td>";
        net += "<td class='text-danger'>" + cd['packets_recv'] + "</td>";
        net += "<td>" + cd['family'] + "</td>";
        net += "<td>" + cd['address'] + "</td>";
        net += "<td>" + cd['netmask'] + "</td>";
        if (cd['broadcast']) {
            net += "<td>" + cd['broadcast'] + "</td></tr>";
        } else {
            net += "<td>无</td></tr>";
        }
    }
    document.getElementById("tb_net").innerHTML = net;

    /*磁盘使用信息实时更新*/
    let disk = "";
    for (let k in data["disk"]) {
        let cd = data["disk"][k];
        disk += "<tr><td>" + cd['device'] + "</td>";
        disk += "<td>" + cd['mountpoint'] + "</td>";
        disk += "<td>" + cd['fstype'] + "</td>";
        disk += "<td>" + cd['opts'] + "</td>";
        disk += "<td class='text-danger'>" + cd['used']['total'] + "GB</td>";
        disk += "<td class='text-danger'>" + cd['used']['used'] + "GB</td>";
        disk += "<td class='text-danger'>" + cd['used']['free'] + "GB</td>";
        disk += "<td><div class='progress'><div class='progress-bar progress-bar-striped progress-bar-animated" + progress_status(cd['used']['percent']) + "' role='progressbar' aria-valuenow='" + cd['used']['percent'] + "' aria-valuemin='0' aria-valuemax='100' style='width: " + cd['used']['percent'] + "%'>" + cd['used']['percent'] + "%</div></div></td></tr>";
    }
    document.getElementById("tb_disk").innerHTML = disk;
}

//2.定义连接函数
function connect() {
    disconnect();//把之前没关闭的连接关闭掉，再创建新的连接
    //创建连接对象
    conn = new WebSocket(`ws://${window.location.host}/real/time/`);
    log("warning", "正在连接....");
    //建立连接
    conn.onopen = function () {
        console.log("连接成功！");
        log("success", "连接成功！");
    };
    //建立接受消息
    conn.onmessage = function (e) {
        //console.log(e.data)
        update_ui(e);
    };
    //建立关闭连接
    conn.onclose = function () {
        console.log("连接断开！");
        log("danger", "连接断开！");
    };
    //每隔几秒触发一个事件
    setInterval(function () {
        conn.send("system");
    }, 500);
}

//3.定义断开连接函数
function disconnect() {
    if (conn != null) {
        log("danger", "连接断开！");
        conn.close(); //关闭连接
        conn = null;
    }
}

//4.刷新页面时候，断开连接，重新连接，断线重连判断
if (conn == null) {
    connect()
} else {
    disconnect();
}
